src/Microsoft.AspNetCore.Razor.Language/CodeGeneration/RuntimeBasicWriter.cs

```diff
@@ -239,42 +240,60 @@ public override void WriteCSharpAttributeValue(CSharpRenderingContext context, C
             {
                 if (node.Children[i] is RazorIRToken token && token.IsHtml)
                 {
-                    builder.Append(token.Content);
+                    var remainingLength = token.Content.Length;
+
+                    while (remainingLength > 0)
+                    {
+                        if (!insideWrite)
+                        {
+                            WriteHtmlContentStartMethodInvocation(context, WriteHtmlContentMethod);
+                            stringLiteralWriter = context.Writer.BuildStringLiteralWriter();
+                            insideWrite = true;
+                        }
+
+                        var lengthToWrite = Math.Min(remainingLength, MaxStringLiteralLength - writtenLength);
+                        stringLiteralWriter.Write(token.Content, token.Content.Length - remainingLength, lengthToWrite);
+
+                        remainingLength -= lengthToWrite;
+                        writtenLength += lengthToWrite;
+
+                        if (writtenLength >= MaxStringLiteralLength)
+                        {
+                            stringLiteralWriter.Dispose();
+                            WriteHtmlContentEndMethodInvocation(context);
+                            insideWrite = false;
+                            writtenLength = 0;
+                        }
+                    }
                 }
             }
 
-            var content = builder.ToString();
-
-            var charactersConsumed = 0;
-
-            // Render the string in pieces to avoid Roslyn OOM exceptions at compile time: https://github.com/aspnet/External/issues/54
-            while (charactersConsumed < content.Length)
+            if (insideWrite)
             {
-                string textToRender;
-                if (content.Length <= MaxStringLiteralLength)
-                {
-                    textToRender = content;
-                }
-                else
-                {
-                    var charactersToSubstring = Math.Min(MaxStringLiteralLength, content.Length - charactersConsumed);
-                    textToRender = content.Substring(charactersConsumed, charactersToSubstring);
-                }
+                stringLiteralWriter.Dispose();
+                WriteHtmlContentEndMethodInvocation(context);
+            }
+        }
 
-                context.Writer
-                    .WriteStartMethodInvocation(WriteHtmlContentMethod)
-                    .WriteStringLiteral(textToRender)
-                    .WriteEndMethodInvocation();
+        protected virtual void WriteHtmlContentStartMethodInvocation(CSharpRenderingContext context, string methodName)
+        {
+            context.Writer.WriteStartMethodInvocation(methodName);
+        }
 
-                charactersConsumed += textToRender.Length;
-            }
+        protected virtual void WriteHtmlContentEndMethodInvocation(CSharpRenderingContext context)
```

Comment:
The `WriteHtmlContentStartMethodInvocation` method is virtual. I find logical to make its associated method `WriteHtmlContentEndMethodInvocation` virtual too, even if it is not currently overriden.